### ISSUE TITLE
common-updater-scripts: incorrect handling of latest nix's hash mismatch messages

### ISSUE THREAD

**Issue by dtzWill** [NixOS/nixpkgs#54962]
## Issue description

When used with latest Nix the "new" hash is extracted incorrectly, `sha256 = "got:   sha256:hashgoeshere` or thereabouts.  This seems to be due to Nix changing the message used to be more friendly to humans:

https://github.com/NixOS/nix/pull/2574

### Steps to reproduce

Use an update script invoking common-updater-scripts.  Sorry, LMK if repro is needed.

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

I don't expect this to be helpful but happy to provide anyway!

 - system: `"x86_64-linux"`
 - host os: `Linux 4.20.5, NixOS, 19.03.git.06b6bb8 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3dtz6739_232d967`
 - channels(will): `"nur-nopkgs, testing, unstable-19.03pre166987.bc41317e243, nixos-1609-16.09.1943.25f4906da6"`
 - channels(root): `"nixos-18.09.1922.97e0d53d669"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by jtojnar** [comment]
```
--- a/pkgs/common-updater/scripts/update-source-version
+++ b/pkgs/common-updater/scripts/update-source-version
@@ -96,7 +96,7 @@ fi
 if [ -z "$newHash" ]; then
     nix-build --no-out-link -A "$attr.src" 2>"$attr.fetchlog" >/dev/null || true
     # FIXME: use nix-build --hash here once https://github.com/NixOS/nix/issues/1172 is fixed
-    newHash=$(egrep -v "killing process|dependencies couldn't be built" "$attr.fetchlog" | tail -n2 | sed "s~output path .* has .* hash ‘\(.*\)’ when .* was expected\|fixed-output derivation produced path '.*' with .* hash '\(.*\)' instead of the expected hash '.*'~\1\2~" | head -n1)
+    newHash=$(egrep -v "killing process|dependencies couldn't be built|wanted: " "$attr.fetchlog" | tail -n2 | sed "s~output path .* has .* hash ‘\(.*\)’ when .* was expected\|fixed-output derivation produced path '.*' with .* hash '\(.*\)' instead of the expected hash '.*'\|  got:    .*:\(.*\)~\1\2\3~" | head -n1)
 fi
 
 if [ -z "$newHash" ]; then

```

It is increasingly messy, perhaps we can drop the old regexes, most people probably set `nix.package` instead of changing nix in overlays so this would not be affected.

**Comment by typetetris** [comment]
What about adding a flag to nix, which turns on machine readable error messages with a guaranty that those messages will not change in a backward incompatible way?
Would need someone to add that.
But wouldn't that be better to parse error messages intended for human readers with shell tools?

**Comment by jtojnar** [comment]
Alternatively, if the build fails, I think we could do something like:

```
nix-hash --flat --base32 --type sha256 $(nix-instantiate -A "$attr.src" --json --eval --strict)
```

**Comment by dtzWill** [comment]
On Thu, 31 Jan 2019 09:21:25 -0800, Jan Tojnar <notifications@github.com> wrote:
> ```
> --- a/pkgs/common-updater/scripts/update-source-version
> +++ b/pkgs/common-updater/scripts/update-source-version
> @@ -96,7 +96,7 @@ fi
>  if [ -z "$newHash" ]; then
>      nix-build --no-out-link -A "$attr.src" 2>"$attr.fetchlog" >/dev/null || true
>      # FIXME: use nix-build --hash here once https://github.com/NixOS/nix/issues/1172 is fixed
> -    newHash=$(egrep -v "killing process|dependencies couldn't be built" "$attr.fetchlog" | tail -n2 | sed "s~output path .* has .* hash ‘\(.*\)’ when .* was expected\|fixed-output derivation produced path '.*' with .* hash '\(.*\)' instead of the expected hash '.*'~\1\2~" | head -n1)
> +    newHash=$(egrep -v "killing process|dependencies couldn't be built|wanted: " "$attr.fetchlog" | tail -n2 | sed "s~output path .* has .* hash ‘\(.*\)’ when .* was expected\|fixed-output derivation produced path '.*' with .* hash '\(.*\)' instead of the expected hash '.*'\|  got:    .*:\(.*\)~\1\2\3~" | head -n1)
>  fi
>  
>  if [ -z "$newHash" ]; then
> 
> ```
> 
> It is increasingly messy, perhaps we can drop the old regexes, most people probably set `nix.package` instead of changing nix in overlays so this would not be affected.

I'm not sure I follow, but FWIW I ran into this and I set `nix.package`
and don't set it via overlay.  Looks like the new behavior is in 2.2,
the current default for `nix`:

```
$ nix run -f channel:nixos-unstable nix -i -c nix eval '(builtins.fetchurl { url = https://nixos.org; sha256 = "1111111111111111111111111111111111111111111111111111"; })'
error: hash mismatch in file downloaded from 'https://nixos.org':
  wanted: sha256:1111111111111111111111111111111111111111111111111111
  got:    sha256:1q7rksf170b1zhy54b5cwig77ycgdiaibcbh47aq8rrw3szra24w
```

(err formatting might not work for email responses, maybe? sorry :))

I think it's reasonable to drop support for nix older than what's in the tree--
especially since you'd have to overlay to replace it and as you say
would likely do via `nix.package` anyway.

Oh... that's probably what you meant! :D

**Comment by dtzWill** [comment]
> `nix-hash --flat --base32 --type sha256 $(nix-instantiate -A "$attr.src" --json --eval --strict)`

Can we do that on a path that isn't valid? I was going to try `nix hash-path` or so but basically they all have the same problem :).  Maybe we should just add the extra sed :innocent:.

**Comment by siraben** [comment]
Any updates on this?

**Comment by FliegendeWurst** [comment]
Should have been fixed by #55605

**Comment by dtzWill** [comment]
Thanks!
